src/os_pcap_nodelet.cpp

```diff
@@ -0,0 +1,252 @@
+        // OusterSensor has its own RingBuffer of PacketMsg but for
+        // now we are focusing on optimizing the code for OusterDriver
+        std::memcpy(imu_packet.buf.data(), raw_imu_packet,
+                    imu_packet.buf.size());
+        imu_packet_pub.publish(imu_packet);
+    }
+
+    void read_packets(PcapReader& pcap, const sensor::packet_format& pf) {
+        size_t payload_size = pcap.next_packet();
+        auto packet_info = pcap.current_info();
+        while (payload_size) {
+            auto start = std::chrono::high_resolution_clock::now();
+            if (packet_info.dst_port == info.config.udp_port_imu) {
+                imu_packets->write_overwrite(
+                    [this, &pcap, &pf, &packet_info](uint8_t* buffer) {
+                        std::memcpy(buffer, pcap.current_data(),
+                                    pf.imu_packet_size);
+                    });
+            } else if (packet_info.dst_port == info.config.udp_port_lidar) {
+                lidar_packets->write_overwrite(
+                    [this, &pcap, &pf, &packet_info](uint8_t* buffer) {
+                        std::memcpy(buffer, pcap.current_data(),
+                                    pf.lidar_packet_size);
+                    });
+            } else {
+                std::cout << "unknown packet" << std::endl;
+            }
+            auto prev_packet_ts = packet_info.timestamp;
+            payload_size = pcap.next_packet();
+            packet_info = pcap.current_info();
+            auto curr_packet_ts = packet_info.timestamp;
+            auto end = std::chrono::high_resolution_clock::now();
+            auto dt = (curr_packet_ts - prev_packet_ts) - (end - start);
+            std::this_thread::sleep_for(dt);  // pace packet generation
+        }
+    }
+
+   private:
+    std::shared_ptr<PcapReader> pcap;
+    PacketMsg lidar_packet;
+    PacketMsg imu_packet;
+    ros::Publisher lidar_packet_pub;
+    ros::Publisher imu_packet_pub;
+    ros::ServiceServer reset_srv;
+
+    std::unique_ptr<ThreadSafeRingBuffer> lidar_packets;
+    std::unique_ptr<ThreadSafeRingBuffer> imu_packets;
+
+    std::atomic<bool> packet_read_active = {false};
+    std::unique_ptr<std::thread> packet_read_thread;
+
+    std::atomic<bool> imu_packets_processing_thread_active = {false};
+    std::unique_ptr<std::thread> imu_packets_processing_thread;
+
+    std::atomic<bool> lidar_packets_processing_thread_active = {false};
+    std::unique_ptr<std::thread> lidar_packets_processing_thread;
+
+    bool force_sensor_reinit = false;
+    bool reset_last_init_id = true;
+
+    bool last_init_id_initialized = false;
+    uint32_t last_init_id;
+
+    // TODO: add as a ros parameter
+    const int max_poll_client_error_count = 10;
+    int poll_client_error_count = 0;
+    // TODO: add as a ros parameter
+    const int max_read_lidar_packet_errors = 60;
+    int read_lidar_packet_errors = 0;
```

Comment:
Shouldnt these be named as ```read_lidar_packet_errors_```?

Reply:
I am split on this since the driver is a ros code, this would be a major shift in the style. I would leave it current as is but we can discuss it offline.

Reply:
Yeah, if everything else is that way I'd leave it for now, but I've always used the trailing underscore style in ROS code before.